android/app/src/main/java/au/com/robin/sms/MainActivity.kt

```diff
@@ -0,0 +1,69 @@
+package au.com.robin.sms
+
+import android.Manifest
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.os.Bundle
+import android.util.Log
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import androidx.core.content.ContextCompat
+import au.com.robin.sms.databinding.ActivityMainBinding
+
+class MainActivity : AppCompatActivity() {
+    private lateinit var smsInterfaceBinding: ActivityMainBinding
+    private val PERMISSION_REQUEST_CODE = 1
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        smsInterfaceBinding = ActivityMainBinding.inflate(layoutInflater)
+        val view = smsInterfaceBinding.root
+        setContentView(view)
+
+        smsInterfaceBinding.btSend.setOnClickListener {
+            val phoneNo = smsInterfaceBinding.etPhoneNo.text?.toString()
+            val message = smsInterfaceBinding.etMessage.text?.toString()
+
+            // Check for permission
+            if (checkPermission()) {
+                // create an intent to let client choose which sim card slot
+                // for now, we use sim card 2. Normally, user will be shown options to choose from.
+                val intent = Intent()
+                intent.putExtra("slot", 1)
+                Log.d("onStart, button", "Sending SMS")
+                val smsManager = SmsSend.getSmsManager(this, intent)
+                smsManager?.sendTextMessage(phoneNo, null, message, null, null)
+            } else {
+                requestPermission()
+            }
+        }
+    }
+
+    private fun checkPermission(): Boolean {
+        val sendSmsPermission =
+            ContextCompat.checkSelfPermission(this, Manifest.permission.SEND_SMS)
+        val readPhoneStatePermission =
+            ContextCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE)
+
+        return sendSmsPermission == PackageManager.PERMISSION_GRANTED && readPhoneStatePermission == PackageManager.PERMISSION_GRANTED
+    }
+
+    private fun requestPermission() {
+        ActivityCompat.requestPermissions(
+            this,
+            arrayOf(Manifest.permission.SEND_SMS, Manifest.permission.READ_PHONE_STATE),
+            PERMISSION_REQUEST_CODE
+        )
+    }
+
+    override fun onRequestPermissionsResult(
+        requestCode: Int,
+        permissions: Array<out String>,
```

Comment:
What does the `out` mean?

Reply:
At first I didn't know that too, as I just tried to follow the doc. But now that you are bringing that up helps me understand it. So it is basically a modifier stating that it is immutable.

Reply:
Ok great. Please link me a page explaining it.

Reply:
Please refer to this https://kotlinlang.org/docs/generics.html#declaration-site-variance